Add gradual typing constraint definitions

migeed-z opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80147
* #80110
* #80102
* #80095
* #80084
* __->__ #79912

Differential Revision: [D37315020](https://our.internmc.facebook.com/intern/diff/D37315020)


Add high-level gradual type constraint definitions. Those definitions will be used by the constraint generator. The generator will generate high-level constraints for FX graphs which will then be transformed to constraints that are understood by Z3

migeed-z committed:
Add gradual typing constraint definitions

[ghstack-poisoned]

migeed-z committed:
Update on "Add gradual typing constraint definitions"




[ghstack-poisoned]

migeed-z committed:
Update on "Add gradual typing constraint definitions"




[ghstack-poisoned]

migeed-z committed:
Update on "Add gradual typing constraint definitions"


Differential Revision: [D37315020](https://our.internmc.facebook.com/intern/diff/D37315020)

[ghstack-poisoned]

migeed-z committed:
Update on "Add gradual typing constraint definitions"


Differential Revision: [D37315020](https://our.internmc.facebook.com/intern/diff/D37315020)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37315020).

migeed-z committed:
Update on "Add gradual typing constraint definitions"


Differential Revision: [D37315020](https://our.internmc.facebook.com/intern/diff/D37315020)

[ghstack-poisoned]

migeed-z committed:
Update on "Add gradual typing constraint definitions"


Differential Revision: [D37315020](https://our.internmc.facebook.com/intern/diff/D37315020)

[ghstack-poisoned]

migeed-z committed:
Update on "Add gradual typing constraint definitions"


Differential Revision: [D37315020](https://our.internmc.facebook.com/intern/diff/D37315020)

[ghstack-poisoned]

anijain2305 commented on torch/fx/experimental/migrate_gradual_types/constraint.py:
```diff
@@ -0,0 +1,388 @@
+
+    def __eq__(self, other):
+        if isinstance(other, Conj):
+            return self.conjucts == other.conjucts and self.conjucts == other.conjucts
```
Just trying to learn - why do you have `self.conjucts == other.conjucts` twice?

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37315020).

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37315020).

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37315020).

jamesr66a reviewed (approved):
Seems fine to me. Do you mind adding a description of what the PR is adding in the description field?

migeed-z committed:
Update on "Add gradual typing constraint definitions"


Differential Revision: [D37315020](https://our.internmc.facebook.com/intern/diff/D37315020)


Add high-level gradual type constraint definitions. Those definitions will be used by the constraint generator. The generator will generate high-level constraints for FX graphs which will then be transformed to constraints that are understood by Z3


[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37315020).

migeed-z committed:
Update on "Add gradual typing constraint definitions"


Differential Revision: [D37315020](https://our.internmc.facebook.com/intern/diff/D37315020)


Add high-level gradual type constraint definitions. Those definitions will be used by the constraint generator. The generator will generate high-level constraints for FX graphs which will then be transformed to constraints that are understood by Z3


[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37315020).

migeed-z commented:
@pytorchbot merge -g
